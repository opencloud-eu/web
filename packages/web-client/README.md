# web-client

The `web-client` is a standalone package that allows you to interact with the [OpenCloud](https://github.com/opencloud-eu/opencloud/) APIs via TypeScript. It provides an abstraction layer between the server and a (web-) application that converts API data into objects with helpful types and utilities. This abstraction ensures that users of the APIs don't need in-depth knowledge about them, such as required methods or returned status codes.

The supported APIs are:

- Graph (drives, sharing, user & group management)
- OCS (capabilities & url signing)
- WebDAV (file operations)

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-client

$ pnpm add @opencloud-eu/web-client

$ yarn add @opencloud-eu/web-client
```

## Usage

### Graph

The graph client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { graph } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const graphClient = graph(baseURI, axiosClient)
```

The following example demonstrates how to retrieve all spaces accessible to the user. A `SpaceResource` can then be used to e.g. fetch files and folders (see webdav example down below).

```
const mySpaces = await graphClient.drives.listMyDrives()
```

### OCS

The ocs client needs to be instantiated with a base URI corresponding to your OpenCloud deployment and an axios instance. The axios instance is being used for all requests, which means it needs to include all relevant headers either statically or via interceptor.

```
import axios from axios
import { ocs } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const axiosClient = axios.create({
	headers: { Authorization: accessToken }
})

const ocsClient = ocs(baseURI, axiosClient)
```

The following examples demonstrate how to fetch capabilities and sign URLs.

```
const capabilities = await ocsClient.getCapabilities()

const signedUrl = await ocsClient.signUrl('some_url_to_sign', 'your_username')
```

### WebDav

The webdav client needs to be instantiated with a base URI corresponding to your OpenCloud deployment. You can also pass a header callback which will be called with every dav request.

```
import { webdav } from '@opencloud-eu/web-client'

const accessToken = 'some_access_token'
const baseURI = 'some_base_uri'

const webDavClient = webdav(
	baseURI,
	() => ({ Authorization: accessToken })
)
```

The following example demonstrates how to list all resources of a given `SpaceResource` (see above how to fetch space resources).

```
const { resource, children } = await webDavClient.listFiles(spaceResource)
```
