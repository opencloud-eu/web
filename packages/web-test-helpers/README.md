# web-test-helpers

This packages provides utilities for unit testing within the OpenCloud app ecosystem.

## Installation

Depending on your package manager, run one of the following commands:

```
$ npm install @opencloud-eu/web-test-helpers --save-dev

$ pnpm add -D @opencloud-eu/web-test-helpers

$ yarn add @opencloud-eu/web-test-helpers --dev
```

Make sure that you have `vue`, `@vue/test-utils`, and `@opencloud-eu/web-pkg` as dev dependencies of your app since those are peer dependencies of this package.
