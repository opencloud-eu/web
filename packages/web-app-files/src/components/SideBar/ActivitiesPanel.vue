<template>
  <oc-loader v-if="isLoading" />
  <template v-else>
    <p v-if="!activities.length" v-text="$gettext('No activities')" />
    <div v-else class="oc-ml-s">
      <oc-list class="oc-timeline">
        <li v-for="activity in activities" :key="activity.id">
          <div class="oc-flex oc-flex-middle">
            <oc-avatars
              :items="getAvatarsFromActivity(activity)"
              class="oc-mr-xs"
              stacked
              gap-size="small"
              :width="16.8"
              icon-size="xsmall"
            >
              <template #userAvatars="{ avatars, width }">
                <user-avatar
                  v-for="avatar in avatars"
                  :key="avatar.userId"
                  :user-id="avatar.userId"
                  :user-name="avatar.userName"
                  :width="width"
                />
              </template>
            </oc-avatars>
            <span v-html="getHtmlFromActivity(activity)" />
          </div>
          <span
            class="oc-text-muted oc-text-small oc-mt-s"
            v-text="getTimeFromActivity(activity)"
          />
        </li>
      </oc-list>
      <p class="oc-text-muted oc-text-small" v-text="activitiesFooterText" />
    </div>
  </template>
</template>

<script setup lang="ts">
import { computed, inject, Ref, ref, unref, watch } from 'vue'
import { useGettext } from 'vue3-gettext'
import { formatDateFromDateTime, useClientService, UserAvatar } from '@opencloud-eu/web-pkg'
import { useTask } from 'vue-concurrency'
import { call, Resource } from '@opencloud-eu/web-client'
import { DateTime } from 'luxon'
import { Activity } from '@opencloud-eu/web-client/graph/generated'
import escape from 'lodash-es/escape'

const { $ngettext, current: currentLanguage } = useGettext()
const clientService = useClientService()
const resource = inject<Ref<Resource>>('resource')
const activities = ref<Activity[]>([])
const activitiesLimit = 200

const activitiesFooterText = computed(() => {
  return $ngettext(
    'Showing %{activitiesCount} activity',
    'Showing %{activitiesCount} activities',
    unref(activities).length,
    {
      activitiesCount: unref(activities).length.toString()
    }
  )
})

const loadActivitiesTask = useTask(function* (signal) {
  activities.value = yield* call(
    clientService.graphAuthenticated.activities.listActivities(
      `itemid:${unref(resource).fileId} AND limit:${activitiesLimit} AND sort:desc`,
      { signal }
    )
  )
}).restartable()

const isLoading = computed(() => {
  return loadActivitiesTask.isRunning || !loadActivitiesTask.last
})

const getHtmlFromActivity = (activity: Activity) => {
  let message = activity.template.message
  for (const [key, value] of Object.entries(activity.template.variables)) {
    const escapedValue = escape(value.displayName || value.name)

    message = message.replace(`{${key}}`, `<strong>${escapedValue}</strong>`)
  }
  return message
}

const getAvatarsFromActivity = (activity: Activity) => {
  const avatars = []

  for (const key of ['user', 'sharee']) {
    const entry = activity.template.variables[key]
    if (entry) {
      avatars.push({
        userName: entry.displayName,
        displayName: entry.displayName,
        userId: entry.id,
        avatarType: entry.shareType || 'user'
      })
    }
  }

  return avatars
}

const getTimeFromActivity = (activity: Activity) => {
  const dateTime = DateTime.fromISO(activity.times.recordedTime)
  return formatDateFromDateTime(dateTime, currentLanguage)
}

watch(
  resource,
  () => {
    loadActivitiesTask.perform()
  },
  {
    immediate: true,
    deep: true
  }
)
</script>
