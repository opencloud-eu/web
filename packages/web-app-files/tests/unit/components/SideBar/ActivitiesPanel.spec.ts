import ActivitiesPanel from '../../../../src/components/SideBar/ActivitiesPanel.vue'
import { defaultComponentMocks, defaultPlugins, mount } from '@opencloud-eu/web-test-helpers'
import { Resource } from '@opencloud-eu/web-client'
import { mock } from 'vitest-mock-extended'
import { flushPromises } from '@vue/test-utils'

const defaultActivities = [
  {
    id: '5380e156-d65e-4024-9691-0f0c1f2796e4',
    times: {
      recordedTime: '2024-07-29T18:34:40Z'
    },
    template: {
      message: '{user} created {resource}.',
      variables: {
        user: {
          id: '71f9de60-8b24-4cfe-9378-87d47aef0d04',
          displayName: 'Marie Curie'
        },
        resource: {
          id: '7f92b0a9-06ad-49dc-890f-0e0a6eb4dea6$e9f01ca3-577f-4d1d-acd4-1cc44149ac25!5fb9f87c-a317-467b-9882-eb9f171564ac',
          name: 'new folder'
        }
      }
    }
  },
  {
    id: '5380e156-d65e-4024-9691-0f0c1f2796e4',
    times: {
      recordedTime: '2023-07-29T18:34:40Z'
    },
    template: {
      message: '{user} moved {resource}.',
      variables: {
        user: {
          id: '71f9de60-8b24-4cfe-9378-87d47aef0d04',
          displayName: 'Albert Einstein'
        },
        resource: {
          id: '7f92b0a9-06ad-49dc-890f-0e0a6eb4dea6$e9f01ca3-577f-4d1d-acd4-1cc44149ac25!5fb9f87c-a317-467b-9882-eb9f171564ac',
          name: 'textfile.txt'
        }
      }
    }
  },
  {
    id: '5380e156-d65e-4024-9691-0f0c1f2796e4',
    times: {
      recordedTime: '2022-07-29T18:34:40Z'
    },
    template: {
      message: '{user} deleted {resource}.',
      variables: {
        user: {
          id: '71f9de60-8b24-4cfe-9378-87d47aef0d04',
          displayName: 'Robert Oppenheimer'
        },
        resource: {
          id: '7f92b0a9-06ad-49dc-890f-0e0a6eb4dea6$e9f01ca3-577f-4d1d-acd4-1cc44149ac25!5fb9f87c-a317-467b-9882-eb9f171564ac',
          name: 'atom plans.pdf'
        }
      }
    }
  },
  {
    id: '5380e156-d65e-4024-9691-0f0c1f2796e4',
    times: {
      recordedTime: '2021-07-29T18:34:40Z'
    },
    template: {
      message: '{user} removed {resource}.',
      variables: {
        user: {
          id: '71f9de60-8b24-4cfe-9378-87d47aef0d04',
          displayName: 'Albert Schweitzer'
        },
        resource: {
          id: '7f92b0a9-06ad-49dc-890f-0e0a6eb4dea6$e9f01ca3-577f-4d1d-acd4-1cc44149ac25!5fb9f87c-a317-467b-9882-eb9f171564ac',
          name: 'Selfie.png'
        }
      }
    }
  }
]
describe('ActivitiesPanel', () => {
  it('should show no activities message if there is no data', async () => {
    const { wrapper } = getMountedWrapper({ activities: [] })
    await flushPromises()
    expect(wrapper.html()).toContain('No activities')
  })
  it('should show loading spinner when fetching data', () => {
    const { wrapper } = getMountedWrapper()
    expect(wrapper.find('.oc-loader').exists()).toBeTruthy()
  })
  it('should render a list of activities when data is present', async () => {
    const { wrapper } = getMountedWrapper()
    await flushPromises()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getMountedWrapper({
  activities = defaultActivities
}: {
  activities?: any[]
  isActive?: boolean
} = {}) {
  const mocks = {
    ...defaultComponentMocks()
  }
  mocks.$clientService.graphAuthenticated.activities.listActivities.mockResolvedValue(activities)

  return {
    wrapper: mount(ActivitiesPanel, {
      global: {
        mocks,
        plugins: [...defaultPlugins()],
        provide: { ...mocks, resource: mock<Resource> }
      }
    }),
    mocks
  }
}
