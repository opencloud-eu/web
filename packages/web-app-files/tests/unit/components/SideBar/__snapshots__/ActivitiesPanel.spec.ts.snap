// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActivitiesPanel > should render a list of activities when data is present 1`] = `
"<div class="oc-ml-s">
  <ul class="oc-list oc-my-rm oc-mx-rm oc-timeline">
    <li>
      <div class="oc-flex"><span class="oc-mr-xs"><span class="oc-avatars oc-avatars-gap-m oc-avatars-stacked" aria-hidden="true"><span class="vue-avatar--wrapper oc-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: #8e5c11; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Marie Curie"><span class="avatar-initials" style="color: white;">MC</span></span>
        <!--v-if-->
        <!--v-if--></span>
        <!--v-if--></span> <span><strong>Marie Curie</strong> created <strong>new folder</strong>.</span>
      </div> <span class="oc-text-muted oc-text-small oc-mt-s">Jul 29, 2024, 6:34 PM</span>
    </li>
    <li>
      <div class="oc-flex"><span class="oc-mr-xs"><span class="oc-avatars oc-avatars-gap-m oc-avatars-stacked" aria-hidden="true"><span class="vue-avatar--wrapper oc-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: #c21c53; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Albert Einstein"><span class="avatar-initials" style="color: white;">AE</span></span>
        <!--v-if-->
        <!--v-if--></span>
        <!--v-if--></span> <span><strong>Albert Einstein</strong> moved <strong>textfile.txt</strong>.</span>
      </div> <span class="oc-text-muted oc-text-small oc-mt-s">Jul 29, 2023, 6:34 PM</span>
    </li>
    <li>
      <div class="oc-flex"><span class="oc-mr-xs"><span class="oc-avatars oc-avatars-gap-m oc-avatars-stacked" aria-hidden="true"><span class="vue-avatar--wrapper oc-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: #3F51B5; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Robert Oppenheimer"><span class="avatar-initials" style="color: white;">RO</span></span>
        <!--v-if-->
        <!--v-if--></span>
        <!--v-if--></span> <span><strong>Robert Oppenheimer</strong> deleted <strong>atom plans.pdf</strong>.</span>
      </div> <span class="oc-text-muted oc-text-small oc-mt-s">Jul 29, 2022, 6:34 PM</span>
    </li>
    <li>
      <div class="oc-flex"><span class="oc-mr-xs"><span class="oc-avatars oc-avatars-gap-m oc-avatars-stacked" aria-hidden="true"><span class="vue-avatar--wrapper oc-avatar" style="width: 16.8px; height: 16.8px; line-height: 16.8px; background-color: #673AB7; font-size: 6px; font-family: Helvetica, Arial, sans-serif; color: white;" width="16.8" aria-hidden="true" focusable="false" data-test-user-name="Albert Schweitzer"><span class="avatar-initials" style="color: white;">AS</span></span>
        <!--v-if-->
        <!--v-if--></span>
        <!--v-if--></span> <span><strong>Albert Schweitzer</strong> removed <strong>Selfie.png</strong>.</span>
      </div> <span class="oc-text-muted oc-text-small oc-mt-s">Jul 29, 2021, 6:34 PM</span>
    </li>
  </ul>
  <p class="oc-text-muted oc-text-small">Showing 4 activities</p>
</div>"
`;
